src/globalPlug.ts

```diff
@@ -0,0 +1,192 @@
+import {
+    EvaluationFacadeOptions as EvaluationOptions,
+    ExternalEvent,
+    ExternalEventPayload,
+    ExternalEventType,
+    JsonValue,
+    Logger,
+    SdkFacade,
+    SessionFacade,
+    TrackerFacade,
+    UserFacade,
+} from '@croct-tech/sdk';
+import {Plug, Configuration} from './plug';
+import {PluginController} from './plugin';
+
+export class GlobalPlug implements Plug {
+    private instance?: SdkFacade;
+
+    private plugins: {[key: string]: PluginController} = {};
+
+    private initialize: {(): void};
+
+    private initialized: Promise<void>;
+
+    public constructor() {
+        this.initialized = new Promise(resolve => {
+            this.initialize = resolve;
+        });
+    }
+
+    public plug(configuration: Configuration): void {
+        if (this.instance !== undefined) {
+            const logger = this.instance.getLogger();
+
+            logger.info('Croct is already plugged in.');
+
+            return;
+        }
+
+        const {plugins, ...sdkConfiguration} = configuration;
+        const sdk = SdkFacade.init(sdkConfiguration);
+
+        this.instance = sdk;
+
+        const logger = this.instance.getLogger();
+        const initialization: Promise<void>[] = [];
+
+        for (const plugin of plugins ?? []) {
+            const pluginName = plugin.getName();
+
+            logger.debug(`Initializing plugin "${pluginName}"...`);
+
+            const controller = plugin.initialize({
+                tracker: sdk.tracker,
+                evaluator: sdk.evaluator,
+                user: sdk.user,
+                session: sdk.session,
+                getLogger: (...namespace: string[]): Logger => sdk.getLogger(pluginName, ...namespace),
+                getTabStorage: (...namespace: string[]): Storage => sdk.getTabStorage(pluginName, ...namespace),
+                getBrowserStorage: (...namespace: string[]): Storage => sdk.getBrowserStorage(pluginName, ...namespace),
+            });
+
+            logger.debug(`Plugin "${pluginName}" initialized`);
+
+            if (typeof controller !== 'object') {
+                continue;
+            }
+
+            this.plugins[pluginName] = controller;
+
+            if (typeof controller.enable === 'function') {
+                const promise = controller.enable();
+
+                if (promise instanceof Promise) {
+                    initialization.push(
+                        promise
+                            .then(() => logger.debug(`Plugin "${pluginName}" enabled`))
+                            .catch(() => logger.error(`Failed to enable plugin "${pluginName}"`)),
+                    );
+                }
```

Comment:
What happens if there is an enable function but it doesn't return a promise?
The signature is `() => Promise<void>|void`, it should be logged either way.

Reply:
Good catch